File: AgentLens/Views/Memory/MemoryReviewInboxModel.swift
import Foundation
import OpenBurnBarCore

/// View-model backing the memory review inbox (`MemoryReviewInboxView`).
///
/// The inbox is the human gate in the quarantine lifecycle (G4): newly extracted
/// memories land as `.quarantined` and cannot be injected or replicated until the
/// user approves them here. This model loads two buckets — pending (quarantined)
/// and approved — opening each sealed body transiently for display only. It never
/// persists or logs the opened body; the string lives only in `Item.body` for the
/// lifetime of the view.
///
/// U7: the inbox serves usage memories (`.safariAsk` / `.agentSession`) alongside
/// chat. Each bucket is loaded once per storage partition — chat with the exact
/// pre-U7 request/paging semantics, usage kinds with the same semantics over the
/// `usage:` partition — and merged in page order, so the chat lane stays
/// byte-identical and neither lane can crowd the other out of its page cap.
/// A second filter axis (`SourceFilter`) narrows the visible rows by source, and
/// every review action routes through the source-kind-guarded store methods with
/// the acted-on row's own kind.
///
/// The closures (`loadPage` / `openBody` / `setStatus` / `forget`) are the seam the
/// integrator wires to the real `MemoryServing` implementation; keeping them injected
/// lets the tests drive the model with in-memory fakes. The init signature and public
/// surface are a frozen contract — coordinate any change with the integrator.
@Observable @MainActor
final class MemoryReviewInboxModel {

    /// One reviewable memory: the authority record plus its transiently-opened body.
    struct Item: Identifiable, Equatable {
        enum BodyLoadState: Equatable {
            case loaded
            case unavailable
        }

        let memory: Memory          // OpenBurnBarCore.Memory
        let body: String            // transiently-opened sealed body (display only)
        let bodyLoadState: BodyLoadState
        var id: MemoryID { memory.id }
        var canApprove: Bool {
            bodyLoadState == .loaded &&
                body.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty == false
        }
    }

    /// Which bucket the inbox is showing.
    enum Filter: String, CaseIterable, Identifiable {
        case pending, approved
        var id: String { rawValue }
        var title: String { self == .pending ? "Pending" : "Approved" }
    }

    /// Second filter axis (U7): which extraction source the rows came from.
    /// `.all` admits chat plus both usage kinds; the source-chip row binds here.
    enum SourceFilter: String, CaseIterable, Identifiable {
        case all, chat, safariAsk, agentSession

        var id: String { rawValue }

        var title: String {
            switch self {
            case .all: "All"
            case .chat: "Chat"
            case .safariAsk: "Safari asks"
            case .agentSession: "Agent sessions"
            }
        }

        /// Kinds this chip admits; `nil` means no restriction.
        var sourceKinds: Set<MemorySourceKind>? {
            switch self {
            case .all: nil
            case .chat: [.chat]
            case .safariAsk: [.safariAsk]
            case .agentSession: [.agentSession]
            }
        }
    }

    typealias LoadPage = (MemoryPageRequest, Set<MemorySourceKind>) async throws -> MemoryPage
    typealias OpenBody = (MemoryID) async throws -> String?
    typealias SetStatus = (MemoryID, MemoryReviewStatus, Set<MemorySourceKind>) async throws -> Bool
    typealias Forget = (MemoryID, Set<MemorySourceKind>) async throws -> Bool

    /// Every source kind the inbox serves. Daemon-owned `.code` rows are not
    /// review-inbox material and stay invisible here.
    static let servedSourceKinds: Set<MemorySourceKind> =
        Set([MemorySourceKind.chat]).union(MemorySourceKind.usageKinds)

    private(set) var pending: [Item] = []
    private(set) var approved: [Item] = []
    private(set) var isLoading = false
    private(set) var errorMessage: String?
    var filter: Filter = .pending
    var sourceFilter: SourceFilter
    var items: [Item] {
        let bucket = filter == .pending ? pending : approved
        guard let kinds = sourceFilter.sourceKinds else { return bucket }
        return bucket.filter { kinds.contains($0.memory.sourceKind) }
    }
    /// Pending across every served source — usage rows count toward the badge.
    var pendingCount: Int { pending.count }

    private let scope: MemoryScope
    private let loadPage: LoadPage
    private let openBody: OpenBody
    private let setStatus: SetStatus
    private let forgetRecord: Forget

    /// Largest page we request per bucket and per source partition. The inbox is
    /// a review surface, not a paginated browser, so a single generous page keeps
    /// the UI simple while still bounding the fetch.
    private let pageSize = 200

    private static let defaultErrorMessage = "Something went wrong loading your memories. Please try again."
    private static let unavailableBodyApprovalMessage = "Memory contents are unavailable. Reject it or reload before approving."

    init(
        scope: MemoryScope,
        sourceFilter: SourceFilter = .all,
        loadPage: @escaping LoadPage,
        openBody: @escaping OpenBody,
        setStatus: @escaping SetStatus,
        forget: @escaping Forget
    ) {
        self.scope = scope
        self.sourceFilter = sourceFilter
        self.loadPage = loadPage
        self.openBody = openBody
        self.setStatus = setStatus
        self.forgetRecord = forget
    }

    /// Loads both buckets: pending keeps only `.quarantined` (requesting quarantined
    /// rows), approved keeps only `.approved`. Each bucket loads chat and usage
    /// partitions separately — the chat call is the exact pre-U7 fetch — and merges
    /// them in page order. Each kept memory's sealed body is opened best-effort for
    /// display. Drives `isLoading` and surfaces failures via `errorMessage`.
    func load() async {
        isLoading = true
        errorMessage = nil
        defer { isLoading = false }

        do {
            let pendingChat = try await loadBucket(
                sourceKinds: [.chat],
                includeQuarantined: true,
                keep: .quarantined
            )
            let pendingUsage = try await loadBucket(
                sourceKinds: MemorySourceKind.usageKinds,
                includeQuarantined: true,
                keep: .quarantined
            )
            let approvedChat = try await loadBucket(
                sourceKinds: [.chat],
                includeQuarantined: false,
                keep: .approved
            )
            let approvedUsage = try await loadBucket(
                sourceKinds: MemorySourceKind.usageKinds,
                includeQuarantined: false,
                keep: .approved
            )
            pending = Self.mergedInPageOrder(pendingChat, pendingUsage)
            approved = Self.mergedInPageOrder(approvedChat, approvedUsage)
        } catch {
            errorMessage = friendlyMessage(for: error)
        }
    }

    /// Approves a memory, then reloads so it moves from pending to approved.
    func approve(_ id: MemoryID) async {
        guard pending.first(where: { $0.id == id })?.canApprove == true else {
            errorMessage = Self.unavailableBodyApprovalMessage
            return
        }
        await transition(id, to: .approved)
    }

    /// Rejects a memory, then reloads so it leaves the pending bucket.
    func reject(_ id: MemoryID) async {
        await transition(id, to: .rejected)
    }

    /// Hard-deletes a memory (member-visible "forget this"): the authority row,
    /// sealed body, and provenance are removed; an approved row leaves a fact
    /// tombstone behind so the deletion replicates (PR1 semantics). Reloads so
    /// the row leaves whichever bucket held it.
    func forget(_ id: MemoryID) async {
        do {
            _ = try await forgetRecord(id, reviewSourceKinds(for: id))
            await load()
        } catch {
            errorMessage = friendlyMessage(for: error)
        }
    }

    // MARK: - Private

    private func transition(_ id: MemoryID, to status: MemoryReviewStatus) async {
        do {
            _ = try await setStatus(id, status, reviewSourceKinds(for: id))
            await load()
        } catch {
            errorMessage = friendlyMessage(for: error)
        }
    }

    /// The source-kind guard for an action on `id`: exactly the acted-on row's
    /// kind, so a usage action can never touch a chat row (or vice versa). Falls
    /// back to every served kind when the row is no longer loaded — the store's
    /// own kind guard still applies, and a stale action then no-ops safely.
    private func reviewSourceKinds(for id: MemoryID) -> Set<MemorySourceKind> {
        (pending + approved)
            .first { $0.id == id }
            .map { [$0.memory.sourceKind] } ?? Self.servedSourceKinds
    }

    /// Merge two per-partition bucket loads back into page order (updatedAt DESC,
    /// id ASC) — the same comparator `memoryPage` serves pages in, so a chat-only
    /// merge is exactly the pre-U7 chat list.
    private static func mergedInPageOrder(_ lhs: [Item], _ rhs: [Item]) -> [Item] {
        (lhs + rhs).sorted { a, b in
            if a.memory.updatedAt == b.memory.updatedAt { return a.id < b.id }
            return a.memory.updatedAt > b.memory.updatedAt
        }
    }

    private func loadBucket(
        sourceKinds: Set<MemorySourceKind>,
        includeQuarantined: Bool,
        keep status: MemoryReviewStatus
    ) async throws -> [Item] {
        var items: [Item] = []
        items.reserveCapacity(pageSize)

        var nextPage = 1
        while items.count < pageSize {
            let request = MemoryPageRequest(
                scope: scope,
                page: nextPage,
                pageSize: pageSize,
                includeQuarantined: includeQuarantined
            )
            let page = try await loadPage(request, sourceKinds)
            guard page.items.isEmpty == false else { break }

            for memory in page.items where memory.reviewStatus == status {
                let bodyState: (String, Item.BodyLoadState)
                do {
                    if let body = try await openBody(memory.id),
                       body.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty == false {
                        bodyState = (body, .loaded)
                    } else {
                        bodyState = ("", .unavailable)
                    }
                } catch {
                    AppLogger.dataStore.silentFailure("Failed to open memory body for \(memory.id)", error: error)
                    bodyState = ("", .unavailable)
                }
                items.append(Item(memory: memory, body: bodyState.0, bodyLoadState: bodyState.1))
                if items.count >= pageSize { break }
            }

            if nextPage * pageSize >= page.total { break }
            nextPage += 1
        }
        return items
    }

    private func friendlyMessage(for error: Error) -> String {
        (error as? LocalizedError)?.errorDescription ?? Self.defaultErrorMessage
    }
}

File: windows/app/OpenBurnBar.App.Presentation/Memory/MemoryReviewInboxModel.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Runtime.CompilerServices;
using System.Threading.Tasks;

namespace OpenBurnBar.App.Presentation.Memories;

// PORTED (faithful) from AgentLens/Views/Memory/MemoryReviewInboxModel.swift.
//
// The inbox is the human gate in the quarantine lifecycle (G4): newly extracted
// memories land as `quarantined` and cannot be injected or replicated until the
// user approves them here. This model loads two buckets — pending (quarantined) and
// approved — opening each sealed body transiently for display only. It never
// persists or logs the opened body; the string lives only in `Item.Body` for the
// lifetime of the view.
//
// The delegates (LoadPage / OpenBody / SetStatus) are the seam the integrator wires
// to the real memory store; keeping them injected lets the tests drive the model
// with in-memory fakes — exactly as the Swift `@Observable @MainActor` original does.
//
// Swift `@Observable` becomes INotifyPropertyChanged here so WinUI x:Bind refreshes.
// Swift `@MainActor` (single-thread affinity) is the caller's responsibility on
// Windows (the view drives it from the UI DispatcherQueue); the logic itself is
// portable and deterministic, which is what the macOS `dotnet test` exercises.

/// <summary>
/// View-model backing the memory review inbox (Windows <c>MemoryReviewInboxView</c>).
/// </summary>
public sealed class MemoryReviewInboxModel : INotifyPropertyChanged
{
    /// <summary>Which bucket the inbox is showing. Swift: <c>enum Filter</c>.</summary>
    public enum Filter
    {
        Pending,
        Approved,
    }

    public delegate Task<MemoryPage> LoadPageDelegate(MemoryPageRequest request);

    public delegate Task<string?> OpenBodyDelegate(string memoryId);

    public delegate Task<bool> SetStatusDelegate(string memoryId, MemoryReviewStatus status);

    /// <summary>Largest page requested per bucket. The inbox is a review surface, not a
    /// paginated browser, so a single generous page bounds the fetch. Swift: <c>pageSize = 200</c>.</summary>
    private const int PageSize = 200;

    private const string DefaultErrorMessage =
        "Something went wrong loading your memories. Please try again.";

    private const string UnavailableBodyApprovalMessage =
        "Memory contents are unavailable. Reject it or reload before approving.";

    private readonly MemoryScope _scope;
    private readonly LoadPageDelegate _loadPage;
    private readonly OpenBodyDelegate _openBody;
    private readonly SetStatusDelegate _setStatus;
    private readonly bool _readOnly;

    private IReadOnlyList<MemoryReviewItem> _pending = Array.Empty<MemoryReviewItem>();
    private IReadOnlyList<MemoryReviewItem> _approved = Array.Empty<MemoryReviewItem>();
    private bool _isLoading;
    private string? _errorMessage;
    private Filter _filter = Filter.Pending;

    public MemoryReviewInboxModel(
        MemoryScope scope,
        LoadPageDelegate loadPage,
        OpenBodyDelegate openBody,
        SetStatusDelegate setStatus,
        bool readOnly = false)
    {
        _scope = scope ?? throw new ArgumentNullException(nameof(scope));
        _loadPage = loadPage ?? throw new ArgumentNullException(nameof(loadPage));
        _openBody = openBody ?? throw new ArgumentNullException(nameof(openBody));
        _setStatus = setStatus ?? throw new ArgumentNullException(nameof(setStatus));
        _readOnly = readOnly;
        if (readOnly)
        {
            _filter = Filter.Approved;
        }
    }

    /// <summary>
    /// Windows v1 is a read mirror of synced approved facts. Approve/reject stays Mac-owned (G4).
    /// </summary>
    public bool IsReadOnly => _readOnly;

    /// <summary>Whether approve/reject actions should be shown.</summary>
    public bool AllowsReviewActions => !_readOnly;

    public string TitleText => _readOnly ? "Synced memories" : "Approve what to remember";

    public string SubtitleText => _readOnly
        ? "Approved facts synced from your Mac appear here. Review and approve still happens on Mac — quarantined facts never leave that device."
        : "Memories stay quarantined until you approve them — nothing is used in a chat before you say so.";

    public string EmptyTitleText => _readOnly ? "No synced memories yet" : "No memories to review";

    public string EmptyBodyText => _readOnly
        ? "Once memory cloud sync is enabled on your Mac and approved facts replicate, they show up here read-only."
        : "When a chat turns up a durable fact or preference worth remembering, it lands here for your approval.";

    public IReadOnlyList<MemoryReviewItem> Pending
    {
        get => _pending;
        private set { _pending = value; OnPropertyChanged(); OnPropertyChanged(nameof(PendingCount)); OnPropertyChanged(nameof(HasPending)); OnPropertyChanged(nameof(PendingPillText)); RaiseDerived(); }
    }

    public IReadOnlyList<MemoryReviewItem> Approved
    {
        get => _approved;
        private set { _approved = value; OnPropertyChanged(); RaiseDerived(); }
    }

    public bool IsLoading
    {
        get => _isLoading;
        private set { if (_isLoading != value) { _isLoading = value; OnPropertyChanged(); OnPropertyChanged(nameof(ShowEmpty)); } }
    }

    public string? ErrorMessage
    {
        get => _errorMessage;
        private set { if (_errorMessage != value) { _errorMessage = value; OnPropertyChanged(); OnPropertyChanged(nameof(HasError)); } }
    }

    public Filter CurrentFilter
    {
        get => _filter;
        set { if (_filter != value) { _filter = value; OnPropertyChanged(); RaiseDerived(); } }
    }

    /// <summary>The bucket currently shown. Swift: <c>items</c>.</summary>
    public IReadOnlyList<MemoryReviewItem> Items => _filter == Filter.Pending ? _pending : _approved;

    /// <summary>Badge count of memories awaiting review. Swift: <c>pendingCount</c>.</summary>
    public int PendingCount => _pending.Count;

    /// <summary>Whether the pending badge/pill should show. Swift: <c>model.pendingCount &gt; 0</c>.</summary>
    public bool HasPending => _pending.Count > 0;

    /// <summary>Pill label, e.g. "3 pending". Swift: the header pending pill text.</summary>
    public string PendingPillText => $"{_pending.Count} pending";

    /// <summary>Whether the current bucket has any rows (drives the list vs empty-state switch).</summary>
    public bool HasItems => Items.Count > 0;

    /// <summary>Show the empty state only when not loading and the bucket is empty.</summary>
    public bool ShowEmpty => !_isLoading && Items.Count == 0;

    /// <summary>Whether an error banner should show.</summary>
    public bool HasError => !string.IsNullOrEmpty(_errorMessage);

    private void RaiseDerived()
    {
        OnPropertyChanged(nameof(Items));
        OnPropertyChanged(nameof(HasItems));
        OnPropertyChanged(nameof(ShowEmpty));
    }

    /// <summary>
    /// Loads both buckets: pending keeps only <c>Quarantined</c> (requesting quarantined
    /// rows), approved keeps only <c>Approved</c>. Each kept memory's sealed body is opened
    /// best-effort for display. Drives <see cref="IsLoading"/> and surfaces failures via
    /// <see cref="ErrorMessage"/>. Swift: <c>func load()</c>.
    /// </summary>
    public async Task LoadAsync()
    {
        IsLoading = true;
        ErrorMessage = null;
        try
        {
            var pendingItems = await LoadBucketAsync(includeQuarantined: true, keep: MemoryReviewStatus.Quarantined).ConfigureAwait(false);
            var approvedItems = await LoadBucketAsync(includeQuarantined: false, keep: MemoryReviewStatus.Approved).ConfigureAwait(false);
            Pending = pendingItems;
            Approved = approvedItems;
        }
        catch (Exception error)
        {
            ErrorMessage = FriendlyMessage(error);
        }
        finally
        {
            IsLoading = false;
        }
    }

    /// <summary>Approves a memory, then reloads so it moves from pending to approved.
    /// Guarded on <c>CanApprove</c> like the Swift original. Swift: <c>func approve(_:)</c>.</summary>
    public async Task ApproveAsync(string id)
    {
        if (_readOnly)
        {
            ErrorMessage = new MemoryReviewMacOnlyException().FriendlyMessage;
            return;
        }

        var item = FindPending(id);
        if (item is null || !item.CanApprove)
        {
            ErrorMessage = UnavailableBodyApprovalMessage;
            return;
        }

        await TransitionAsync(id, MemoryReviewStatus.Approved).ConfigureAwait(false);
    }

    /// <summary>Rejects a memory, then reloads so it leaves the pending bucket.
    /// Swift: <c>func reject(_:)</c>.</summary>
    public Task RejectAsync(string id)
    {
        if (_readOnly)
        {
            ErrorMessage = new MemoryReviewMacOnlyException().FriendlyMessage;
            return Task.CompletedTask;
        }

        return TransitionAsync(id, MemoryReviewStatus.Rejected);
    }

    // MARK: - Private

    private MemoryReviewItem? FindPending(string id)
    {
        foreach (var item in _pending)
        {
            if (string.Equals(item.Id, id, StringComparison.Ordinal))
            {
                return item;
            }
        }

        return null;
    }

    private async Task TransitionAsync(string id, MemoryReviewStatus status)
    {
        try
        {
            _ = await _setStatus(id, status).ConfigureAwait(false);
            await LoadAsync().ConfigureAwait(false);
        }
        catch (Exception error)
        {
            ErrorMessage = FriendlyMessage(error);
        }
    }

    private async Task<IReadOnlyList<MemoryReviewItem>> LoadBucketAsync(bool includeQuarantined, MemoryReviewStatus keep)
    {
        var items = new List<MemoryReviewItem>(PageSize);
        int nextPage = 1;

        while (items.Count < PageSize)
        {
            var request = new MemoryPageRequest(
                Scope: _scope,
                Page: nextPage,
                PageSize: PageSize,
                IncludeQuarantined: includeQuarantined);

            var page = await _loadPage(request).ConfigureAwait(false);
            if (page.Items.Count == 0)
            {
                break;
            }

            foreach (var memory in page.Items)
            {
                if (memory.ReviewStatus != keep)
                {
                    continue;
                }

                string body;
                MemoryBodyLoadState state;
                try
                {
                    var opened = await _openBody(memory.Id).ConfigureAwait(false);
                    if (!string.IsNullOrWhiteSpace(opened))
                    {
                        body = opened!;
                        state = MemoryBodyLoadState.Loaded;
                    }
                    else
                    {
                        body = string.Empty;
                        state = MemoryBodyLoadState.Unavailable;
                    }
                }
                catch
                {
                    body = string.Empty;
                    state = MemoryBodyLoadState.Unavailable;
                }

                items.Add(new MemoryReviewItem(memory, body, state, ShowReviewActions: !_readOnly));
                if (items.Count >= PageSize)
                {
                    break;
                }
            }

            if (nextPage * PageSize >= page.Total)
            {
                break;
            }

            nextPage += 1;
        }

        return items;
    }

    private static string FriendlyMessage(Exception error) =>
        (error as IFriendlyError)?.FriendlyMessage ?? DefaultErrorMessage;

    public event PropertyChangedEventHandler? PropertyChanged;

    private void OnPropertyChanged([CallerMemberName] string? name = null) =>
        PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(name));
}

/// <summary>Whether the sealed body opened for display. Swift: <c>Item.BodyLoadState</c>.</summary>
public enum MemoryBodyLoadState
{
    Loaded,
    Unavailable,
}

/// <summary>
/// One reviewable memory: the authority record plus its transiently-opened body.
/// Swift: <c>MemoryReviewInboxModel.Item</c> (top-level here so the WinUI DataTemplate
/// can name it in <c>x:DataType</c>). Body-derived display helpers live here so both
/// the model logic and the XAML row bind the same rules.
/// </summary>
public sealed record MemoryReviewItem(
    Memory Memory,
    string Body,
    MemoryBodyLoadState BodyLoadState,
    bool ShowReviewActions = true)
{
    public string Id => Memory.Id;

    /// <summary>Approvable only when the sealed body actually opened to non-empty text —
    /// mirrors the Swift <c>canApprove</c> guard that blocks approving an unavailable body.</summary>
    public bool CanApprove =>
        ShowReviewActions &&
        BodyLoadState == MemoryBodyLoadState.Loaded &&
        !string.IsNullOrWhiteSpace(Body);

    /// <summary>Row body text with the Swift fallback for an unavailable seal.
    /// Swift: <c>MemoryReviewRow.bodyText</c>.</summary>
    public string DisplayBody
    {
        get
        {
            string trimmed = Body.Trim();
            return trimmed.Length == 0 ? "(Memory contents unavailable)" : trimmed;
        }
    }

    /// <summary>Uppercase kind badge text. Swift: <c>item.memory.kind.rawValue</c> (uppercased).</summary>
    public string KindBadge => Memory.Kind.RawValue().ToUpperInvariant();

    /// <summary>Confidence as a whole-percent string. Swift: <c>confidencePercent</c>.</summary>
    public string ConfidenceLabel => $"{(int)System.Math.Round(Memory.Confidence * 100)}%";
}
